trademapper/js/trademapper.mapper.js

```diff
@@ -108,6 +104,116 @@ define(["d3", "topojson", "worldmap", "disputedareas", "countrycentre", "config"
 				.attr("fill", "url(#diagonalHatch)");
 	},
 
+	// load overlays from the overlays= variable in the query string
+	addOverlays: function () {
+		var params = (new URL(document.location)).searchParams;
+		var urls = params.get('overlays');
+
+		if (urls === null) {
+			return;
+		}
+
+		urls = urls.split(',');
+		console.log('FETCHING OVERLAYS FROM URLS:', urls);
+
+		urls.forEach(this.loadTopojsonFromURL.bind(this));
+	},
+
+	loadTopojsonFromURL: function(url) {
+		var self = this;
+
+		d3.json(url).then(
+			this.loadTopojson.bind(this),
+
+			function (error) {
+				console.log("unable to download", error);
+			}
+		);
+	},
+
+	loadTopojson: function (data) {
+		// lines need drawing with class overlay-line;
+		// polygons need drawing with class overlay-area, but also need to
+		// add lines denoting their boundaries;
+		// points can be treated the same way as polygons and don't need boundaries
+		// to be derived
+		var polygons = {
+			type: "GeometryCollection",
+			geometries: []
+		};
+
+		var lines = {
+			type: "GeometryCollection",
+			geometries: []
+		};
+
+		var points = {
+			type: "GeometryCollection",
+			geometries: []
+		};
+
+		// sort the geometries so they can be drawn appropriately
+		var geometries, i, geometry;
+		for (var propertyName in data.objects) {
+			geometries = data.objects[propertyName].geometries;
+
+			for (i = 0; i < geometries.length; i++) {
+				var geometry = geometries[i];
+
+				if (geometry.type === "Polygon" || geometry.type === "MultiPolygon") {
+					polygons.geometries.push(geometry);
+				} else if (geometry.type === "LineString" || geometry.type === "MultiLineString") {
+					lines.geometries.push(geometry);
+				} else if (geometry.type === "Point" || geometry.type === "MultiPoint") {
+					points.geometries.push(geometry);
+				}
+			}
+		}
+
+		if (polygons.geometries.length > 0) {
+			var polygonFeatures = topojson.feature(data, polygons).features;
+
+			this.mapg.selectAll(".overlay-polygon")
+				.data(polygonFeatures)
+				.enter()
+					.append("path")
```

Comment:
Is the additional indent on purpose following the 'enter' ?

Reply:
It's part of the d3 style of writing stuff, to mark scope: enter creates a context for each element selected by selectAll(). You can see it in other parts of the code. I'm not sure I like it; I might change it and assign a variable to the selected elements, then call enter() on that variable, just to clarify what's happening.

Reply:
Or maybe I'll just unindent it, as that's what some code I've seen does (including stuff written by the writer of d3).

Reply:
I decided to leave this, to keep it consistent with other parts of the code which use indents to show the scope of d3 API calls.